master/internal/experiment.go

```diff
@@ -255,45 +210,36 @@ func restoreExperiment(master *Master, expModel *model.Experiment) error {
 		return errors.Wrapf(err, "failed to create experiment %d from model", expModel.ID)
 	}
 
-	log := log.WithField("experiment", e.ID)
-
-	log.Info("restoring experiment")
-	e.replaying = true
-
-	ref, _ := master.system.ActorOf(actor.Addr("experiments", e.ID), e)
-
-	// Wait for the experiment to handle any initial searcher operations.
-	master.system.Ask(ref, doneProcessingSearcherOperations{}).Get()
-
-	if err = e.db.RollbackSearcherEvents(e.ID); err != nil {
-		return errors.Wrapf(err, "failed to rollback searcher events")
-	}
-
-	if err = e.db.ForEachSearcherEvent(e.ID, newSearcherEventCallback(master, ref)); err != nil {
-		return errors.Wrapf(err, "failed to get searcher events")
-	}
-
-	// We have the experiment ask all the trials to restore (since we don't know all of the trial
-	// actor children) and wait here for them to finish. Since the trials might ask things of the
-	// experiment while restoring, we can't have the experiment itself wait for the trials.
-	trialResponses := master.system.Ask(ref, restoreTrials{}).Get()
-
-	// If the experiment failed during the replay we may receive a nil response.
-	if trialResponses == nil {
-		return errors.Errorf("experiment %v did not respond to 'restoreTrials' message", e.ID)
-	}
-
-	for range trialResponses.(actor.Responses) {
+	// We restore the experiment here (and somewhere similar W.R.T. restoring trials) and not
+	// in actor.PreStart or elsewhere in the experiment lifecycle to sidestep needing to reason
+	// about what to do when messages are received in the restoring state or in the running state
+	// (or others).
```

Comment:
Why not establish a pattern where `newExperiment` takes a snapshot (or nil), as does `newTrial`?  Conceptually, that makes a lot of sense.  "An actor is created with its old snapshot so that before any actor messages happen it is right back in the state it was in before".

Reply:
yeah, that's fine with me. this basically is that quote and functionally is the same. we always restore prior to calling ActorOf(...), but I'll see if that pattern looks clearer.